### PR TITLE
Added a check for the files in the ESH-INF dir

### LINKED ISSUE #65: Create check that validates files in ESH-INF directory
The check will validate the .xml files against the ESH thing description XSD, config description XSD and binding XSD.

It will include as well checks for missing or unused:
- config-description.uri ;
- bridge-type-ref.id ;

See: http://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html.

### PR DESCRIPTION
**Pull request by svilenvul**
Closes #65

Signed-off-by: Svilen Valkanov <svilen.valkanov@musala.com>

### CODE REVIEW

**Comment by svilenvul** [comment]
@martinvw, it would be great if you could take a look at this PR.

**Comment by martinvw** [comment]
Try to do it later, maybe in the train on the way home :-)

Can you fix the conflict?

**Comment by svilenvul** [comment]
Yes, for sure.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    private void checkBuildProperties() {
+        if (buildPropertiesFile != null) {
+            IBuildEntry binIncludes = buildPropertiesFile.getEntry(IBuildEntry.BIN_INCLUDES);
+            if (binIncludes != null) {
+                String[] includedTokens = binIncludes.getTokens();
+                // Exclude processed files that are included
+                for (String included : includedTokens) {
+                    // Iterator is used, as the collection will be modified
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: 👍

File: src/main/resources/rulesets/checkstyle/suppressions.xml [hunk 2]
```diff
@@ -8,4 +8,7 @@
     <suppress files=".+\\internal\\.+\.java" checks="JavadocType|JavadocVariable|JavadocMethod|JavadocFilterCheck"/>
     <suppress files=".+DTO\.java" checks="JavadocType|JavadocVariable|JavadocMethod|JavadocFilterCheck" />
     <suppress files=".+Impl\.java" checks="JavadocType|JavadocVariable|JavadocMethod|JavadocFilterCheck"/>
-</suppressions>
\ No newline at end of file
+    <!-- Homematic and Tellstick bindings are creating and configuring things dynamically, they will log false positives for unused configuration -->
+    <!-- IO and Voice bundles have specific use cases - they use only the config .xml files and will log false positives as well -->
+    <suppress files=".+org.openhab.binding.homematic.+|.+org.openhab.binding.tellstick.+|.+org.openhab.io.+|.+org.openhab.voice.+" checks="EshInfXmlCheck"/>
```
Comment (martinvw) [inline src/main/resources/rulesets/checkstyle/suppressions.xml]: Thanks for explaining this, up to @kaikreuzer whether this is okay.

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    private static final String CONFIG_SCHEMA_URL = SCHEMA_ROOT_URL + "config-description-1.0.0.xsd";
+
+    private static final String MESSAGE_EMPTY_FILE = "The file {0} should not be empty.";
+    private static final String MESSAGE_MISSING_URI_CONFIGURATION = "Missing configuration for the configuration reference with uri - {0}";
+    private static final String MESSAGE_MISSING_SUPPORTED_BRIDGE = "Missing the supported bridge with id {0}";
+    private static final String MESSAGE_UNUSED_URI_CONFIGURATION = "Unused configuration reference with uri - {0}";
+    private static final String MESSAGE_UNUSED_BRIDGE = "Unused bridge reference with id - {0}";
+    private static final String MESSAGE_NOT_INCLUDED_XML_FILE = "The file {0} isn`t included in the build.properties file. Good approach is to add include all files by adding `ESH-INF/` value to the bin.includes property.";
+
+    private static final DefaultConfiguration config = createCheckConfig(EshInfXmlCheck.class);
```
Comment (martinvw) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlCheckTest.java]: Sounds like constant, so uppercase

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        allConfigDescriptionRefs = new HashMap<>();
+        allConfigDescriptions = new HashMap<>();
+
+        allSupportedBridges = new HashMap<>();
+        allBridgeTypes = new HashMap<>();
+    }
+
+    @Override
+    public void beginProcessing(String charset) {
+        logger.debug("Executing the {}", this.getClass().getSimpleName());
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: I think it's not important in this case, in both cases they will be displayed only when maven -X command line option is used.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    @Override
+    protected void processFiltered(File file, List<String> lines) throws CheckstyleException {
+        String fileName = file.getName();
+        logger.debug("Processing the {}", fileName);
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Same as above.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java (hunk 4 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Will X also show trace? I already saw mentions that the output exploded when using the -X, so that why mentioned it. Maybe then just leave it out?

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+                if (BUILD_PROPERTIES_FILE_NAME.equals(file.getName())) {
+                    processBuildProperties(file);
+                }
+        }
+    }
+
+    @Override
+    public void finishProcessing() {
+        // Check for missing supported bridge-type-refs.
+        Map<String, File> missingSupportedBridges = intersect(allSupportedBridges, allBridgeTypes);
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Yes, you are completely right of course :), the name is misleading.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java (hunk 4 above)
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: I think the exploding log was caused mainly by this [issue ](https://github.com/checkstyle/checkstyle/issues/3184). Maybe it is time to update checkstyle, 7.6.1 is released :)

**Comment by martinvw** [comment]
Do you also need to update the readme?

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+                        Map<String, File> configDescriptionRefs = evaluateExpressionOnFile(xmlFile,
+                                CONFIG_DESCRIPTION_REF_EXPRESSION);
+                        allConfigDescriptionRefs.putAll(configDescriptionRefs);
+                        Map<String, File> configDescriptions = evaluateExpressionOnFile(xmlFile,
+                                CONFIG_DESCRIPTION_EXPRESSION);
+                        allConfigDescriptions.putAll(configDescriptions);
+                        break;
+                    }
+                    case BINDING_DIRECTORY: {
+                        schema = bindingSchema;
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Yes, the files are different. If they should be configurable, maybe not.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+        }
+        return collection;
+    }
+
+    private NodeList getNodes(File xmlFile, String expression) throws CheckstyleException {
+        Document document = parseDomDocumentFromFile(xmlFile);
+
+        XPathExpression xpathExpression = compileXPathExpression(expression);
+
+        NodeList nodes = null;
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Thanks, but it looks like I can't easily instantiate an empty NodeList.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    private <K> void logMissingEntries(Map<K, File> collection, String message) {
+        for (K element : collection.keySet()) {
+            File xmlFile = collection.get(element);
+            logMessage(xmlFile.getPath(), 0, xmlFile.getName(), MessageFormat.format(message, element));
+        }
+    }
+
+    private void logMessage(String filePath, int line, String fileName, String message) {
+        MessageDispatcher dispatcher = getMessageDispatcher();
+        dispatcher.fireFileStarted(filePath);
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Maybe yes, it is used to instead of log(..) in the finishProcessing() method, I think it's definitely worth to add it in the abstract class and document it.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java (hunk 8 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Pity, the solutions I can find are worse :-(

http://stackoverflow.com/questions/30897704/initialize-nodelist-java#answer-30897953

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java (hunk 9 above)
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: I have found it in three checks - RequiredFilesCheck, ServiceComponentManifestCheck and this one.
I prefer to make a separate PR for this method and include some common constants in the base class (extensions, properties, etc..)

Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: 👍

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlCheckTest.java [hunk 10]
```diff
@@ -0,0 +1,267 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
```
Comment (svilenvul) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlCheckTest.java]: The maven build won't check the headers. I will change it manually.

**Comment by svilenvul** [comment]
> Do you also need to update the readme?
I think it is not part of the guidelines.

**Comment by martinvw** [comment]
But should you maybe update (or any other rule which is (also) covered by this change):

> Every bundle must contain a build.properties file, which lists all resources that should end up in the binary under bin.includes. - Work in Progress

See also your own remark at (https://github.com/openhab/static-code-analysis/pull/52#issuecomment-291848552)

> Great, thanks, please update the README.MD file with the priority of the check.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    @Override
+    protected void processFiltered(File file, List<String> lines) throws CheckstyleException {
+        String fileName = file.getName();
+        logger.debug("Processing the {}", fileName);
+
+        switch (FilenameUtils.getExtension(fileName)) {
+            case THING_TYPE_EXTENSION:
+                processXmlFile(file);
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: I prefer to leave it the way, that is done now.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+                validator.validate(new StreamSource(xmlFile));
+
+            } catch (SAXParseException exception) {
+                String message = exception.getMessage();
+                // Removing the type of the logged message (For example - "cvc-complex-type.2.4.b: ...").
+                message = message.substring(message.indexOf(":") + 2);
+                int lineNumber = exception.getLineNumber();
+                log(lineNumber, message, xmlFile.getPath());
+            } catch (IOException | SAXException e) {
+                logger.error("Problem occurred while parsing the file " + xmlFile.getName(), e);
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: I think no.

**Comment by svilenvul** [comment]
I will do it in https://github.com/openhab/static-code-analysis/pull/74.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+                Path relativePath = bundlePath.relativize(filePath);
+                eshInfFiles.put(relativePath, xmlFile);
+
+                String schema = null;
+                switch (fileParentDirectory.getName()) {
+                    case THING_DIRECTORY: {
+                        schema = thingSchema;
+                        // Process the files for all nodes below,
+                        // the allowed values are described in the thing description XSD
+                        Map<String, File> supportedBridgeTypes = evaluateExpressionOnFile(xmlFile,
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Good point, but should this not be defined in the XSD files :

> Occurrence Indicators
Occurrence indicators are used to define how often an element can occur.
Note: For all "Order" and "Group" indicators (any, all, choice, sequence, group name, and group reference) the default value for maxOccurs and minOccurs is 1. 

See https://www.w3schools.com/xml/schema_complex_indicators.asp.

From what I have read I am left with the impression that we shouldn't write logic to check for duplicates and leave that to the XML validation. - http://stackoverflow.com/questions/26341791/duplicates-in-xml.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+ * Check if all files from ESH-INF are included in the build.properties file.
+ *
+ * @author Aleksandar Kovachev - Initial implementation
+ * @author Svlien Valkanov - Some code refactoring and cleanup, added check for the build.properties file
+ *
+ */
+public class EshInfXmlCheck extends AbstractStaticCheck {
+
+    private static final String THING_TYPE_EXTENSION = "xml";
+    private static final String PROPERTIES_EXTENSTION = "properties";
```
Comment (svilenvul) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: Will be addressed in separate PR, as discussed below.

File: src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java (hunk 13 above)
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java]: No the problem was not duplication in a single file but the same channel being defined in the different xml files.

**Comment by svilenvul** [comment]
Many thanks for the review @martinvw !
